### DISCUSSION TITLE
custom build databases

### DISCUSSION THREAD

**Discussion by bhartimittal2976** [milaboratory/mixcr#1340] [Q&A] [answered]
Hello, I am working on different animal species eg cow, pig, rabbit etc. I would like to use their reference databases. I would like to get training on custom databases. How to get gene feature file for the respective species. I have the fasta files for each regions, Can I know how to get gene feature files. Please contact me at- bharti.mittal@pirbright.ac.uk

Looking forward to hearing,
Regards,
Bharti

**Comment by mizraelson** [comment, marked as answer]
Hi,

We have a dedicated command for creating a custom reference library. You can read the instructions [here](https://mixcr.com/mixcr/guides/create-custom-library/#de-novo-libraries).

Briefly, to create a custom reference library, run the command below:

```
mixcr buildLibrary --debug \
  --v-genes-from-fasta v-genes.IGH.fasta \
  --v-gene-feature VRegion \
  --j-genes-from-fasta j-genes.IGH.fasta \
  --d-genes-from-fasta d-genes.IGH.fasta \ # optional
  --c-genes-from-fasta c-genes.IGH.fasta \ # optional
  --chain IGH \
  --taxon-id 9913 \
  --species cow \
  cow-IGH.json.gz
```

In this command:

- FASTQ files contain genes in regular FASTA format. For instance:
```
>IGHV12-348
GATGCTGGAGTTATCCAGTCACCCCGCCATGAGGTGACAGAGATGGG...
```

- `--v-gene-feature` defines which part of the V gene is provided in the FASTA file. Common values include:
  - `VTranscript` if your sequence contains the 5'UTR, Leader sequences, and the actual gene sequence from FR1.
  - `VRegion` if your FASTA files contain V gene sequences starting from FR1.
  
Simply put, `VTranscript` equates to `V5UTR+L1+L2+VRegion`.

If you're working with multiple chains (e.g., IGH, IGK, IGL), each chain should be processed separately. You can then merge them together using the command below:

```bash
mixcr mergeLibrary \
    cow-IGH.json.gz \
    cow-IGK.json.gz \
    cow-IGL.json.gz \
    cow.json.gz 
```

Subsequently, you can run MiXCR with your custom library using a command similar to the one below. Ensure you select the appropriate command based on your library preparation protocol. Also, make sure your library file (e.g., cow.json.gz) is located either in the same directory from which you're running MiXCR or in one of the directories indicated in the `mixcr -v` output.

```
mixcr analyze generic-amplicon \
    --library cow \
    --species cow \
    --rna \
    --rigid-left-alignment-boundary \
    --floating-right-alignment-boundary C \
    input_R1.fastq.gz \
    input_R2.fastq.gz \
    output
```

I hope this clarifies the process for you. Let me know if you need further assistance.

Sincerely,
Mark

**Reply by bhartimittal2976** [reply]
could you share the format of --v-gene-feature VRegion file format

**Reply by mizraelson** [reply]
`--v-gene-feature VRegion` is not a file but a parameter that defines the feature you have in your fasta file. In this case the sequences should cover everything from FR1 to the end of the V gene

**Comment by bhartimittal2976** [comment]
Thank you.
Bharti
